frame/collective/src/lib.rs

```diff
@@ -185,29 +235,76 @@ decl_module! {
+		///   - where `N` new-members-count (governance-bounded)
+		///   - where `P` proposals-count
+		/// - DB:
+		///   - 1 storage mutation (codec `O(M)` read, `O(N)` write) for reading and writing the members
+		///   - 1 storage read (codec `O(P)`) + `P` storage mutations for updating the votes for each proposal (codec `O(M)`)
+		///   - 2 storage writes (codec `O(1)`) for deleting the old `prime` and setting the new one
+		/// - Benchmark: 0 + M * 28.94 + N * 0 + P * 26.74 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::set_members(T::DbWeight::get(), *old_count, new_members.len() as Weight, T::MaxProposals::get()),
+			DispatchClass::Operational
+		)]
+		fn set_members(origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>, old_count: u32) {
 			ensure_root(origin)?;
+			ensure!(new_members.len() <= T::MaxMembers::get() as usize, Error::<T, I>::TooManyMembers);
+			let old = Members::<T, I>::get();
+			ensure!(old.len() <= old_count as usize, Error::<T, I>::InvalidCount);
 			let mut new_members = new_members;
 			new_members.sort();
-			let old = Members::<T, I>::get();
 			<Self as ChangeMembers<T::AccountId>>::set_members_sorted(&new_members[..], &old);
 			Prime::<T, I>::set(prime);
 		}
 
 		/// Dispatch a proposal from a member using the `Member` origin.
 		///
 		/// Origin must be a member of the collective.
-		#[weight = (100_000_000, DispatchClass::Operational)]
-		fn execute(origin, proposal: Box<<T as Trait<I>>::Proposal>) {
+		///
+		/// # <weight>
+		/// ## Weight
+		/// - `O(M + P)` where `M` members-count and `P` complexity of dispatching `proposal`
+		/// - DB: 1 read (codec `O(M)`) and 1 event + DB access of `proposal`
+		/// - Benchmark: 85.71 + M * 0.326 + B * 0.005 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::execute(T::DbWeight::get(), T::MaxMembers::get(), proposal.get_dispatch_info().weight),
+			DispatchClass::Operational
+		)]
+		fn execute(origin, proposal: Box<<T as Trait<I>>::Proposal>) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			ensure!(Self::is_member(&who), Error::<T, I>::NotMember);
+			let members = Self::members();
+			ensure!(members.contains(&who), Error::<T, I>::NotMember);
 
 			let proposal_hash = T::Hashing::hash_of(&proposal);
-			let ok = proposal.dispatch(RawOrigin::Member(who).into()).is_ok();
-			Self::deposit_event(RawEvent::MemberExecuted(proposal_hash, ok));
+			let result = proposal.dispatch(RawOrigin::Member(who).into());
+			Self::deposit_event(RawEvent::MemberExecuted(proposal_hash, result.is_ok()));
+
+			match result {
+				Ok(post_info) => {
+					Ok(post_info.actual_weight.map(|w| weight_for::execute(
+						T::DbWeight::get(),
+						members.len() as Weight,
+						w
+					)).into())
+				},
+				Err(err) => {
+					Ok(err.post_info.actual_weight.map(|w| weight_for::execute(
+						T::DbWeight::get(),
+						members.len() as Weight,
+						w
+					)).into())
+				}
+			}
```

Comment:
@gavofyork This preserves the previous semantics of `execute`. Is it intentional that the proposal failing to execute is not passed out of `execute`?